esmvaltool/diag_scripts/weighting/climwip.py

```diff
@@ -371,12 +372,71 @@ def compute_overall_mean(dataset: 'xr.Dataset',
     return overall_mean
 
 
+def combine_dimension(dataset: 'xr.DataArray', dimn: str) -> 'xr.DataArray':
+    """Combine ensemble members of the same model along the given dimension."""
+    model_ensemble = dataset[dimn].values
+    models = [me.split('_')[0] for me in model_ensemble]
+    dataset['model'] = xr.DataArray(models, dims=dimn)
+    dataset = dataset.groupby('model').mean(keep_attrs=True).rename(
+        {'model': dimn})
+
+    # new model names
+    groups = {
+        k: [*v]
+        for k, v in itertools.groupby(model_ensemble,
+                                      lambda x: x.split('_')[0])
+    }
+    for key, group in groups.items():
+        model, forcing = group[0].split('_')[::2]
+        members = '-'.join([me.split('_')[1] for me in group])
+        groups[key] = '_'.join([model, members, forcing])
+
+    # ensure that they are sorted correctly
+    return dataset.assign_coords(
+        {dimn: [groups[key] for key in dataset[dimn].values]})
+
+
+def combine_ensemble_members(
+        dataset: Union['xr.DataArray', None]) -> Union['xr.DataArray', None]:
+    """Combine ensemble members of the same model along all dimensions."""
+    if dataset is None:
+        return None
+    dataset = combine_dimension(dataset, 'model_ensemble')
+    if 'perfect_model_ensemble' in dataset.dims:
+        dataset = combine_dimension(dataset, 'perfect_model_ensemble')
+        # need to set the diagonal elements back to zero after averaging
+        dataset.values[np.diag_indices(dataset['model_ensemble'].size)] = 0
+    return dataset
+
+
+def split_ensemble_members(dataset: 'xr.DataArray') -> 'xr.DataArray':
+    """Split combined ensemble members of the same model."""
+    model_ensemble = []
+    nr_members = []
+    for model in dataset['model_ensemble'].data:
+        ensembles = model.split('_')[1].split('-')
+        nr_members.append(len(ensembles))
+        model_ensemble += [
+            '_'.join([model.split('_')[0], ensemble,
+                      model.split('_')[2]]) for ensemble in ensembles
+        ]
+
+    data_scaled = dataset.values / nr_members
+    data_expanded = np.repeat(data_scaled, nr_members)
+
+    return xr.DataArray(data_expanded,
+                        coords={'model_ensemble': model_ensemble},
+                        dims='model_ensemble',
+                        name=dataset.name,
+                        attrs=dataset.attrs)
+
+
 def calculate_weights(
-        performance: Union['xr.DataArray', None],
-        independence: Union['xr.DataArray', None],
+        performance: Union['xr.DataArray',
+                           None], independence: Union['xr.DataArray', None],
```

Comment:
I found the previous code formatting better.

Reply:
I agree but tell that to the pre-commt processor it keeps chaning it and I got tired of fixing it every time after I run it :p But now I fixed it again